olive/engine/packaging/packaging_generator.py

```diff
@@ -32,20 +40,124 @@
+            )
+        elif packaging_config.type == PackagingType.AzureMLModels:
+            _generate_azureml_resources(
+                packaging_config.name,
+                packaging_config.config,
+                footprints,
+                pf_footprints,
+                azureml_client_config,
+                PackagingType.AzureMLModels,
+            )
+        elif packaging_config.type == PackagingType.AzureMLData:
+            _generate_azureml_resources(
+                packaging_config.name,
+                packaging_config.config,
+                footprints,
+                pf_footprints,
+                azureml_client_config,
+                PackagingType.AzureMLData,
+            )
+
+
+def _generate_azureml_resources(
+    output_name: str,
+    config: CommonPackagingConfig,
+    footprints: Dict["AcceleratorSpec", "Footprint"],
+    pf_footprints: Dict["AcceleratorSpec", "Footprint"],
+    azureml_client_config: "AzureMLClientConfig",
+    packaging_type: PackagingType,
+):
+    logger.info("Packaging output models to %s", packaging_type)
+
+    for accelerator_spec, pf_footprint in pf_footprints.items():
+        if pf_footprint.nodes and footprints[accelerator_spec].nodes:
+            model_rank = 1
+            for model_id in pf_footprint.nodes:
+                model_name = f"{output_name}_{accelerator_spec}_{model_rank}"
+                export_in_mlflow_format = config.export_in_mlflow_format
+                with tempfile.TemporaryDirectory() as temp_dir:
+                    tempdir = Path(temp_dir)
+                    inference_config = pf_footprint.get_model_inference_config(model_id) or {}
+
+                    saved_model_path = _save_model(
+                        pf_footprint, model_id, tempdir, inference_config, export_in_mlflow_format
+                    )
+
+                    if packaging_type == PackagingType.AzureMLModels:
+                        _upload_to_azureml_models(azureml_client_config, saved_model_path, model_name, config)
+                    elif packaging_type == PackagingType.AzureMLData:
+                        _upload_to_azureml_data(azureml_client_config, saved_model_path, model_name, config)
+
+                model_rank += 1
+
+
+def _upload_to_azureml_models(
+    azureml_client_config, model_path: Path, model_name: str, config: AzureMLModelsPackagingConfig
+):
+    from azure.ai.ml.constants import AssetTypes
+    from azure.ai.ml.entities import Model
+    from azure.core.exceptions import ServiceResponseError
+
+    ml_client = azureml_client_config.create_client()
+    file_model = Model(
+        path=model_path,
+        type=AssetTypes.MLFLOW_MODEL if config.export_in_mlflow_format else AssetTypes.CUSTOM_MODEL,
+        name=model_name,
+        version=str(config.version),
+        description=config.description,
+    )
+    retry_func(
+        ml_client.models.create_or_update,
```

Comment:
need we wait the rest api result? I am supposing it is async op?

Reply:
This is a [sync api](https://learn.microsoft.com/en-us/python/api/azure-ai-ml/azure.ai.ml.mlclient?view=azure-python#azure-ai-ml-mlclient-create-or-update). There is another one for [async](https://learn.microsoft.com/en-us/python/api/azure-ai-ml/azure.ai.ml.mlclient?view=azure-python#azure-ai-ml-mlclient-begin-create-or-update)